### PR TITLE
Hide Columns

### PR DESCRIPTION
**Pull request by KevinWu098**
## Summary
Columns can now be shown/hidden through a dropdown menu in the Course Render Pane

Demo Hide Column feature

### Columns
- Columns can now be hidden by deselecting them in a new Dropdown Menu housed in `CoursePaneButtonRow`
  - Active columns are stored and emitted in `RightPaneStore.ts`

### Removing ConditionalWrapper
- `ConditionalWrapper.tsx` has been removed in favor of plain conditional rendering logic.
- >Just use conditional rendering logic instead of a goofy component. JSX is already a pain to read, there's no need to obscure component logic even more.

### Removing colorAndDelete
- The property `colorAndDelete`, used in SectionTableBody (and related files) has been removed in favor of more reasonably calculating the proper value for `addedCourse` in `sectionTableBody.tsx`

Previously, `sectionTableBody.tsx` would always pass `colorAndDelete` as false, but then immediately recalculate it with the addedCourse state when it rendered:

```typescript
const SectionTableBody = withStyles(styles)((props: SectionTableBodyProps) => {
    const { classes, section, courseDetails, term, 
      colorAndDelete /*ALWAYS false in courseRenderPane*/, 
      allowHighlight, scheduleNames } = props;

    const [addedCourse, setAddedCourse] = useState(colorAndDelete);

{...}

    // Immediately resets addedCourse
    useEffect(() => {
        const updateHighlight = () => {
            const doAdd = AppStore.getAddedSectionCodes().has(`${section.sectionCode} ${term}`);
            setAddedCourse(doAdd);
        };
```

The new callback doesn't immediately setAddedCourse, so instead we should just set the default useState to:
```typescript
const [addedCourse, setAddedCourse] = useState(
        AppStore.getAddedSectionCodes().has(`${section.sectionCode} ${term}`)
    );
```

Doing so eliminates the need for the `colorAndDelete` prop!

### Index.ts changes
- The way the app is rendered has been changed in `index.tsx` to an async function

### Refactor scheduleConflict
- Refactored and added new helper functions into `calendarizeHelpers.ts` to assist in calculating scheduleConflicts
- Utilize useMemo and useCallback to more correctly manage state

## Test Plan
- [ ] Columns should correctly be derendered when they're deselected in the Menu
- [ ] Make sure it functions well on all devices

## Issues
Zotcourse parity a la #440

### CODE REVIEW

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx [hunk 1]
```diff
@@ -24,14 +27,52 @@ const styles: Styles<Theme, object> = {
...
+    { value: 'location', label: 'Places' },
+    { value: 'sectionEnrollment', label: 'Enrollment' },
+    { value: 'restrictions', label: 'Restrictions' },
+    { value: 'status', label: 'Status' },
+];
+
 interface CoursePaneButtonRowProps {
     classes: ClassNameMap;
     showSearch: boolean;
+    columnSelector: boolean;
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx]: I would rename this to `displayColumnsSelector`

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx [hunk 2]
```diff
@@ -48,6 +89,36 @@ class CoursePaneButtonRow extends PureComponent<CoursePaneButtonRowProps> {
...
+                                <MenuItem key={column.value} value={column.value} style={{ maxWidth: '200px' }}>
+                                    <Checkbox
+                                        checked={this.state.activeColumns.indexOf(column.value) >= 0}
+                                        color="default"
+                                    />
+                                    <ListItemText primary={column.label} />
+                                </MenuItem>
+                            ))}
+                        </Select>
+                    </FormControl>
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx]: I would extract this to be a separate component so that it has its own name so that its function would be immediately obvious

File: apps/antalmanac/src/components/RightPane/RightPaneStore.ts [hunk 3]
```diff
@@ -19,6 +19,18 @@ const defaultFormValues: Record<string, string> = {
     room: '',
 };
 
+// If there's a best practice that doesn't use 'let', LMK!
+const activeColumns = [
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/RightPaneStore.ts]: This doesn't appear to be used

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx [hunk 4]
```diff
@@ -48,6 +89,36 @@ class CoursePaneButtonRow extends PureComponent<CoursePaneButtonRowProps> {
...
+                        <MoreVert />
+                    </IconButton>
+                </Tooltip>
+
+                {this.state.columnSelector && (
+                    <FormControl className={classes.formControl}>
+                        <Select
+                            style={{ width: '0', zIndex: -1, color: isDarkMode() ? '#303030' : '#FAFAFA' }}
+                            inputProps={{ IconComponent: () => null }} // some styling nonsense to Thanos snap away the little arrow icon: https://aguidehub.com/blog/2023-02-25-how-to-hide-arrow-from-mui-select-in-react-js/
+                            open={this.state.columnSelector}
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx]: Wouldn't this always be true because, otherwise, it wouldn't be rendered?

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx [hunk 5]
```diff
@@ -24,14 +27,52 @@ const styles: Styles<Theme, object> = {
...
     showSearch: boolean;
+    columnSelector: boolean;
     onDismissSearchResults: () => void;
     onRefreshSearch: () => void;
 }
 
 class CoursePaneButtonRow extends PureComponent<CoursePaneButtonRowProps> {
+    state = {
+        columnSelector: false,
+        activeColumns: [
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx]: I would make column names an enum and make this an enum array so that all the naming is consistent. You can then translate it in `handleChange`. Alternatively, you can have an interface of `Record<ColumnName, Boolean>`.

Also, I don't get why the column names can't be the name that is displayed

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx [hunk 6]
```diff
@@ -64,12 +66,52 @@ const styles = {
...
+    { value: 'status', label: 'Status' },
+];
+
 const SectionTable = (props: SectionTableProps) => {
     const { classes, courseDetails, term, colorAndDelete, highlightAdded, scheduleNames, analyticsCategory } = props;
     const courseId = courseDetails.deptCode.replaceAll(' ', '') + courseDetails.courseNumber;
     const encodedDept = encodeURIComponent(courseDetails.deptCode);
     const isMobileScreen = useMediaQuery(`(max-width: ${MOBILE_BREAKPOINT}`);
 
+    const [columns, setColumns] = useState([
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx]: I think that having a consolidated mapping of the selected columns would be better.

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx [hunk 7]
```diff
@@ -64,12 +66,52 @@ const styles = {
     },
 };
 
+const TableHeadColumns: { value: string; label: string }[] = [
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx]: This should be camelcase

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx [hunk 8]
```diff
@@ -64,12 +66,52 @@ const styles = {
...
+            setColumns(columns);
+            setActiveTableHeadColumns(TableHeadColumns.filter((column) => columns.includes(column.value)));
+        });
+
+        return () => {
+            RightPaneStore.removeListener('columnChange', (columns) => {
+                setColumns(columns);
+                setActiveTableHeadColumns(TableHeadColumns.filter((column) => columns.includes(column.value)));
+            });
+        };
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx]: Extract callback to `changeColumns()` or something along these lines so that you're not duplicating code

**Review by MinhxNguyen7 — DISMISSED** [review]
Sorry for taking so long to review. I missed the notification. 

Great feature and works well. There are only some code quality issues.

My main issue is that there's so much separate state management and duplication going on. It's difficult to parse logic that should be relatively straightforward. Instead of having slightly different and separate states, try to put it all in one place, have named types, and translate it at the very end when you use it if really necessary.

Also, remember to merge. (I know it's my fault that there's a commit on `main`)

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx (hunk 4 above)
Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx]: you damn right

File: apps/antalmanac/src/components/RightPane/RightPaneStore.ts (hunk 3 above)
Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/RightPaneStore.ts]: you also is damn right here

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx (hunk 6 above)
Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/SectionTable/SectionTable.tsx]: Hopefully resolved by me removing that nonsense activeTableColumn code and instead just filtering during render

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx (hunk 5 above)
Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx]: I mentioned this on Disc, but lemme try to reword it a bit better. I'm not entirely sure how to implement this because my current logic is as follows:

1. `CoursePaneButtonRow` is updated, emitting the new set of activeColumns to `SectionTable` & `SectionTableBody`. 
2. Both SectionTables then translate the activeColumns based on what they're rendering (Header strings vs Components)

Because there's slight variations, (CPBR has Restrictions but ST has Rstr, STB is components not strings), I'm not sure how to transfer the states through RightPaneStore without an in-between through value

Screenshot 2023-06-29 at 10 13 44 PM

File: apps/antalmanac/src/components/AppBar/CustomAppBar.tsx [hunk 9]
```diff
@@ -38,6 +37,14 @@ interface CustomAppBarProps {
     classes: ClassNameMap;
 }
 
+const components = [
+    <SettingsMenu key="settings" />,
+    <ImportStudyList key="studylist" />,
+    <Feedback key="feedback" />,
+    <News key="news" />,
+    <AboutPage key="about" />,
+];
```
Comment (ap0nia) [inline apps/antalmanac/src/components/AppBar/CustomAppBar.tsx:46]: Initializing an array of components is most consistent when done outside a component -- i.e. just like this, especially when concerning function components. Because of React's rendering paradigm, initializing components within components in non-standard ways can mess up state badly (another React footgun :tm:)

File: apps/antalmanac/src/components/ConditionalWrapper.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/ConditionalWrapper.tsx:1]: Good riddance to this abomination of a component. :sob: 💢

File: apps/antalmanac/src/components/PatchNotes.tsx [hunk 10]
```diff
@@ -31,29 +31,27 @@ const PatchNotes = () => {
             >
                 <DialogTitle>{"What's New - May 2023"}</DialogTitle>
                 <DialogContent>
-                    <DialogContentText>
```
Comment (ap0nia) [inline apps/antalmanac/src/components/PatchNotes.tsx]: This component is a `p` tag, don't nest all this random stuff into a `p` tag.

File: apps/antalmanac/src/components/AppBar/CustomAppBar.tsx [hunk 11]
```diff
@@ -60,36 +67,18 @@ const CustomAppBar = ({ classes }: CustomAppBarProps) => {
...
-                    wrapper={(children) => (
-                        <div>
-                            <MenuIcon onClick={handleClick} />
-                            <Menu anchorEl={anchorEl} keepMounted open={Boolean(anchorEl)} onClose={handleClose}>
-                                {children}
-                            </Menu>
-                        </div>
-                    )}
-                >
+                {isMobileScreen ? (
```
Comment (ap0nia) [inline apps/antalmanac/src/components/AppBar/CustomAppBar.tsx:70]: Just use conditional rendering logic instead of a goofy component. JSX is already a pain to read, there's no need to obscure component logic even more.

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx [hunk 12]
```diff
@@ -1,56 +1,161 @@
...
+        [RightPaneStore.setActiveColumns]
+    );
+
+    const handleOpen = useCallback(() => {
+        setOpen(true);
+    }, [setOpen]);
+
+    const handleClose = useCallback(() => {
+        setOpen(false);
+    }, [setOpen]);
```
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneButtonRow.tsx:86]: Technically all callback functions should be memo-ized if you want to write "idiomatic React", but the performance improvements are negligible if you consider that React is one of the worst performant frameworks in existence. I'll just leave these callbacks here and let inconsistencies exist in the codebase where callbacks haven't been memo-ized :shrug:

File: apps/antalmanac/src/index.tsx [hunk 13]
```diff
@@ -1,3 +1,23 @@
-import { render } from 'react-dom';
+import { createRoot } from 'react-dom/client';
 import App from './App';
-render(<App />, document.getElementById('root'));
+
+/**
+ * This function runs in the browser to load the React single-page application.
+ */
+async function main() {
```
Comment (ap0nia) [inline apps/antalmanac/src/index.tsx:7]: I prefer explicitly creating and invoking asynchronous functions for executable scripts because Promises can be scoped without any issues and it's easy to tell what the file wants to accomplish, i.e. run a function when it's imported.

**Review by ap0nia — APPROVED** [review]
I think the changes look fine. I'll wait for the preview deployment to finish before merging.

**Comment by EricPedley** [comment]
Bumping this PR, seems like all that's left is to fix some merge conflicts

**Review by EricPedley — APPROVED** [review]

**Comment by ap0nia** [comment]
wait, stop

**Comment by ap0nia** [comment]
What are you doing, don't do this
```ts
  useEffect(() => {
        const updateHighlight = useCallback(() => {
          setAddedCourse(AppStore.getAddedSectionCodes().has(`${section.sectionCode} ${term}`));
        }, [setAddedCourse, AppStore.getAddedSectionCodes]);

        const checkAndDisplayScheduleConflict = () => {
            if (AppStore.getCourseEventsInCalendar().length < 1) {
                setScheduleConflict(false);
                return;
            }
```

**Comment by ap0nia** [comment]
Don't invoke hooks inside of other hooks

**Comment by KevinWu098** [comment]
💀 I did a whoopsie

**Comment by ap0nia** [comment]
Ok, fine, give me a bit to fix it

**Comment by KevinWu098** [comment]
Sorry 'bout all that nonsense, hopefully this latest commit should resolve the issues, but I'll test on the staging instance // read over the code again right now!

**Comment by ap0nia** [comment]
Ok no like. The entire strategy for figuring out conflicts is just wrong. There shouldn't even be a useEffect + useState. The entire logic should be a single useMemo

File: apps/antalmanac/src/stores/calendarizeHelpers.ts [hunk 14]
```diff
@@ -144,8 +144,8 @@ export const calendarizeCustomEvents = (currentCustomEvents: RepeatingCustomEven
  * @returns the start and end time of a course in a 24 hour time with a leading zero (##:##)
  * @returns undefined if there is no WebSOC time (e.g. 'TBA', undefined)
  */
-export const translateWebSOCTimeTo24HourTime = (section: AASection) => {
-    const timeString = section.meetings[0].time.replace(/\s/g, '');
+export const translateWebSOCTimeTo24HourTime = (time: string) => {
```
Comment (ap0nia) [inline apps/antalmanac/src/stores/calendarizeHelpers.ts]: Can anybody confirm if changing this interface breaks any part of the app? 

The rationale for the API change is ***don't pass in the entire object  to the function if all it needs is a string*** to do its work

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx [hunk 15]
```diff
@@ -349,97 +362,119 @@ interface SectionTableBodyProps {
...
+    const sectionDetails = useMemo(() => {
+        return {
+            // If there already exists a more well-written way to translate secton.meetings days (string) into a number, LMK
+            // Converts SuTuTh -> [Su, Tu, Th] -> [0, 2, 4]
+            day: section.meetings[0].days.match(/[A-Z][a-z]*/g)?.map((day: string) => DAYS_TO_NUMS[day]),
+            translatedTime: translateWebSOCTimeTo24HourTime(section.meetings[0].time),
+            startTime: '',
+            endTime: '',
+        };
+    }, [section.meetings]);
```
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx]: Instead of recalculating this arbitrarily, we know that it should be updated whenever `section.meetings` changes, so memoize it with that

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx [hunk 16]
```diff
@@ -349,97 +362,119 @@ interface SectionTableBodyProps {
...
-            checkAndDisplayScheduleConflict();
+    useEffect(() => {
+        AppStore.on('addedCoursesChange', updateHighlight);
+        AppStore.on('currentScheduleIndexChange', updateHighlight);
+
+        return () => {
+            AppStore.removeListener('addedCoursesChange', updateHighlight);
+            AppStore.removeListener('currentScheduleIndexChange', updateHighlight);
         };
+    }, [updateHighlight]);
```
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx]: For clarity, all event handlers are attached and removed in their own useEffect blocks

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx [hunk 17]
```diff
@@ -349,97 +362,119 @@ interface SectionTableBodyProps {
...
         return () => {
-            AppStore.removeListener('addedCoursesChange', updateCourseState);
-            AppStore.removeListener('currentScheduleIndexChange', updateCourseState);
+            AppStore.removeListener('addedCoursesChange', updateCalendarEvents);
+            AppStore.removeListener('currentScheduleIndexChange', updateCalendarEvents);
         };
-    }, [section.sectionCode, term]); //should only run once on first render since these shouldn't change.
+    }, [updateCalendarEvents]);
+
+    const scheduleConflict = useMemo(() => {
```
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx:438]: This value is derived from other stateful values, so it shouldn't be recorded with `useState` and `useEffect`. If it's declared as an un-memoized value, then it's re-calculated on every single render anyways. It's memoized because that's literally the correct way to write this dumb shit.

File: apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx [hunk 18]
```diff
@@ -19,6 +22,16 @@ import { mobileContext } from '$components/MobileHome';
 import locationIds from '$lib/location_ids';
 import { translateWebSOCTimeTo24HourTime } from '$stores/calendarizeHelpers';
 
+const DAYS_TO_NUMS: { [key: string]: number } = {
```
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/SectionTable/SectionTableBody.tsx]: This is moved outside the function component because it's a const value and we don't really need it re-calculated on every render.

**Review by ap0nia — APPROVED** [review]
Ok, code is fixed. I have to go so I can't immediately verify the behavior live. Somebody else please review :pray:

TLDR:
Please don't write bad code. And if you don't know React, try consulting somebody that does before approving React code.

**Comment by KevinWu098** [comment]
> Deployed staging instance to https://staging-624.antalmanac.com

I'll eventually move this into the original PR body, but for now, this is why I removed colorAndDelete:

Previously, `sectionTableBody.tsx` would always pass `colorAndDelete` as false, but then immediately recalculate it with the addedCourse state when it rendered:

```typescript
const SectionTableBody = withStyles(styles)((props: SectionTableBodyProps) => {
    const { classes, section, courseDetails, term, 
      colorAndDelete /*ALWAYS false in courseRenderPane*/, 
      allowHighlight, scheduleNames } = props;

    const [addedCourse, setAddedCourse] = useState(colorAndDelete);

{...}

    // Immediately resets addedCourse
    useEffect(() => {
        const updateHighlight = () => {
            const doAdd = AppStore.getAddedSectionCodes().has(`${section.sectionCode} ${term}`);
            setAddedCourse(doAdd);
        };
```

The new callback doesn't immediately setAddedCourse, so instead we should just set the default useState to:
```typescript
const [addedCourse, setAddedCourse] = useState(
        AppStore.getAddedSectionCodes().has(`${section.sectionCode} ${term}`)
    );
```

Doing so eliminates the need for the `colorAndDelete` prop!

**Review by ap0nia — APPROVED** [review]
Ok good enough, if it breaks prod I'll fix it later. :rocket:
